lib/ansible/module_utils/service.py

```diff
@@ -215,9 +215,10 @@ def daemonize(module, cmd):
                 for out in list(fds):
                     if out in rfd:
                         data = os.read(out.fileno(), chunk)
-                        if not data:
+                        if data:
+                            output[out] += to_bytes(data, errors=errors)
+                        else:
```

Comment:
So I still think we are incorrectly handling the if and the else/break

```
if (rfd + wfd + efd) or p.poll() is not None:
....
      if data:
              output[out] += to_bytes(data, errors=errors)
      fds.remove(out) ```
We always remove the fd as we either already read it or it was empty, this will force the exit of the while loop in the end and we can remove the `else/break`. Also we don't go through loop prematurely when `is None` and the process is still running. So we wait for it to finish OR the file descriptors to be read.

Reply:
@bcoca ok, let's break it down into 2 parts:

1. In https://github.com/ansible/ansible/pull/81581#discussion_r1306186833 you mentioned `fds.remove(out)` should be under `else`. Are you suggesting otherwise here?
2. I can examine this further, but for some reason this is not working as expected. [This](https://github.com/ansible/ansible-modules-core/blob/devel/system/service.py#L249-L282) is the reference to the old code that's working and what it seems was ported incorrectly in #21256 as I mentioned in #81064. Do you see any misalignment here?

Reply:
#1 yes, but that was in context of keeping the loop as you changed it, I realized that both the previous version and your update still didn't  seem to be correct 

#2  that does seem correct, closer to what i propose, the fds don't exist AND the program has ended ... it is clearer than what we do and did in that ported code.

Reply:
comparing closely, ammended by this PR .. I think the code does align with the 'original original', so this fix might be enough